### PR TITLE
PWGJE: First draft of exploratory track skim

### PR DESCRIPTION
**Pull request by raymondEhlers**
This is the first draft of a minimal track skim approach to access what can be done feasibility. The current plan is get this running at a small scale, see what can be done to make it feasible, and then look to contribute from there. It's a small step, but we hope an entry point towards further involvement

### CODE REVIEW

**Comment by nzardosh** [comment]
Hi Raymond,

This is very nice but it seems to be the same skim and strategy that was reported a few weeks ago by me and jochen and which is currently in development (the part for charged tracks and collisions in data and MC is done and we are working on EMCAL and HF additions before releasing it). If I am misunderstanding the purpose let me know but if not should we wait till that one is committed (should be a few weeks) and then we can review if it needs any additions, just so we avoid having multiple skims from the JE side?

Thanks,
Nima

**Comment by raymondEhlers** [comment]
Hi Nima,

Thanks for your comment! I think our aim here is a little different, but I could have misunderstood your and Jochen's developments - I'm just working off the slides and minutes, so there's definitely lots of space for misunderstanding! My understanding is that you and Jochen were working on an approach to convert AO2D -> slimmed down track skim that would have enough information for everyone to run jet analyses on, but that didn't reference the original AO2D. Is that right?

Our goal here was to try to skim down even further, looking at how minimal we can go and how well we can work with these on local, smaller computing resources. Explicit in this would be trading reduced flexibility for reduced storage. I know this has been discussed in the past, but since this approach has worked well for Berkeley folks, I think starting from a familiar approach could be a good entry point for us to be better involved. It ultimately may not be the way that we proceed, but I think this is one of those instances where even a few test trees could be helpful. I labeled as the track skim as LBL since it's experimental at this point, but of course am happy to collaborate/share/drop/whatever as approriate if the goals align. My plan was that this could modified to run on the skims that are produced by you and Jochen's developments, but based on my understand of O2 functionality, I'm not aware of any other way to eg. reduce precision without writing a separate task.

All that being said, please correct me if I've gone wrong somewhere! Thanks and have a good weekend!

Best,
Raymond

**Comment by nzardosh** [comment]
Hi Raymond, maybe we could set up a meeting early next week to discuss further on zoom. I think the idea is basically the same as we also only keep what is needed in order to be able to run a self contained jet finder, so the goal should be reasonably similar. Its probably best to avoid two parallel developments of the same thing and also to avoid down the line running essentially the same skims twice over the full AOD. If the skims we produce have too much info for you you can of course skim them further to run locally but i guess its best to have one skim over the full AOD that is inclusive of both use cases. I would be happy to meet for example on Tuesday to take a look together?

**Comment by raymondEhlers** [comment]
Sure, let's meet and figure out a plan. I'm at the INT workshop this week, but I'm sure we can find some time. Let's try to sort it on mattermost

Thanks, Raymond

**Review by nzardosh — DISMISSED** [review]

**Comment by raymondEhlers** [comment]
@nzardosh When you have a chance, could you please review/approve? All changes + compiler + bug fixes should be in now. Thanks!

**Comment by nzardosh** [comment]
Hi Raymond, thanks for the updates! In the current configuration what is the difference between your task and the jetderiveddataoutputwriter? I am adding now switches in the writer to select which tables (like BCs or clusters) that one really wants to save, so the only difference in this case would be that you have slightly fewer fields per table and that you also do a very particular track or event selection up front instead of storing a bit? As we discussed before I am very happy to approve this for testing but due to the similarities it might be hard to justify running it over the full sample. However I guess your task can run over the JE derived data sets without a problem and you can then store this even further reduced output starting from those? This could be quite a nice test and if small enough something anyone wanting to do local analysis could use. For this use case I would not forsee a problem if the output sizes are good but it would require all dependency on the original AOD tables to be dropped. Would that be a good approach in your mind? probably its what you intended anyways with the PR?

**Comment by raymondEhlers** [comment]
Hi Nima,

Thanks for your comment - a few responses:

>  slightly fewer fields per table 

I think this is underselling this exploratory skim:

1. Dropping a factor of 2-3x or more in fields is a significant difference in storage for local analysis
2. We can use more compressed fields (which mostly seemed to be done in the derived model), but also to encode information in bitpacking, etc. This may be less critical than the first point

> However I guess your task can run over the JE derived data sets without a problem and you can then store this even further reduced output starting from those

Yes, that's always what I've had in mind

> it would require all dependency on the original AOD tables to be dropped

Unless I've overlooked something, I don't think there is any dependence on the AOD tables anymore? If I've missed something, can you please let me know? Or is there somehow an additional issue with calling `globalIndex()`? If so, how would you suggest that we determine unique indices for runs, tracks, etc? (to be clear, I understand this provides a link to other tables, but I'm not so worried about that - I just want a unique identifier that we could trace back if needed for further investigations)

Best,
Raymond

**Comment by nzardosh** [comment]
Hi Raymond,

The problem is that you have an index to Collisions instead of JCollisions as I commented in the PR. I think if this is fixed then it would be independent of the original AO2D and would work well.

Thanks,
Nima

**Comment by raymondEhlers** [comment]
Hi Nima,

Sorry, I'm quite confused - I think your review on the PR was revised and addressed quite a while ago. I just rebased to squash commits, but I made changes to adapt to the new framework (including Collision -> JCollision) a couple of weeks ago. Currently in the code, I see:

```
[10:28] REhlers@REhlers alice/O2Physics/PWGJE/Tasks $ git grep "aod::JCollision" trackSkimExploratory.cxx
trackSkimExploratory.cxx:    soa::Filtered<aod::JCollisions>::iterator const& collision,
trackSkimExploratory.cxx:    soa::Join<aod::JCollisions, aod::JMcCollisionLbs>::iterator const& collision, soa::Join<aod::JTracks, aod::JMcTrackLbs> const& tracks,
trackSkimExploratory.cxx:    soa::Filtered<aod::JCollisions>::iterator const& collision,
[10:28] REhlers@REhler alice/O2Physics/PWGJE/Tasks $ git grep "aod::Collision" trackSkimExploratory.cxx
```

Can you please highlight the particular line where you see the issue? I don't mean to be difficult - I just don't see it

Thanks,
Raymond

**Review by nzardosh — APPROVED** [review]

File: PWGJE/Tasks/trackSkimExploratory.cxx [hunk 1]
```diff
@@ -0,0 +1,349 @@
...
+using namespace o2::framework;
+using namespace o2::framework::expressions;
+
+// Define track skim table
+namespace o2::aod
+{
+namespace ExploratorySkim
+{
+// NOTE: We don't want a Collision Index ID, as this will require the full input AO2D, which we don't want!
+DECLARE_SOA_INDEX_COLUMN(Collision, collision);
```
[the remark this answers is not in the document]
Comment (raymondEhlers) [reply PWGJE/Tasks/trackSkimExploratory.cxx]: Thanks for pointing this out - it's fixed now. Sorry to have overlooked this - I think you might have posted the commit as part of a review, but the overall review may not have gotten posted until we chatted. In any case, should be resolved now, and thanks for your help!

**Pull request closed without merging**